// oc/dist/registry/domain/validators/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateVersion = exports.validateRegistryConfiguration = exports.validatePluginsRequirements = exports.validatePackageJson = exports.validatePackage = exports.validateOcCliVersion = exports.validateNodeVersion = exports.validateComponentParameters = exports.validateComponentName = void 0;
const semver_1 = __importDefault(require("semver"));
const oc_cli_version_1 = __importDefault(require("./oc-cli-version"));
const component_parameters_1 = __importDefault(require("./component-parameters"));
const package_json_validator_1 = __importDefault(require("./package-json-validator"));
const plugins_requirements_1 = __importDefault(require("./plugins-requirements"));
const registry_configuration_1 = __importDefault(require("./registry-configuration"));
const uploaded_package_1 = __importDefault(require("./uploaded-package"));
const node_version_1 = __importDefault(require("./node-version"));
function validateComponentName(componentName) {
    return !/[^a-zA-Z0-9\-_]/.test(componentName) && componentName !== '_package';
}
exports.validateComponentName = validateComponentName;
exports.validateComponentParameters = component_parameters_1.default;
exports.validateNodeVersion = node_version_1.default;
exports.validateOcCliVersion = oc_cli_version_1.default;
exports.validatePackage = uploaded_package_1.default;
exports.validatePackageJson = package_json_validator_1.default;
exports.validatePluginsRequirements = plugins_requirements_1.default;
exports.validateRegistryConfiguration = registry_configuration_1.default;
function validateVersion(version) {
    return !!semver_1.default.valid(version);
}
exports.validateVersion = validateVersion;
