// oc/dist/registry/middleware/request-handler.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const response_time_1 = __importDefault(require("response-time"));
const events_handler_1 = __importDefault(require("../domain/events-handler"));
function requestHandler() {
    return (0, response_time_1.default)((req, res, time) => {
        const data = {
            body: req.body,
            duration: parseInt(String(time * 1000)),
            headers: req.headers,
            method: req.method,
            path: req.path,
            relativeUrl: req.originalUrl,
            query: req.query,
            url: req.protocol + '://' + req.get('host') + req.originalUrl,
            statusCode: res.statusCode
        };
        if (res.errorDetails) {
            data.errorDetails = res.errorDetails;
        }
        if (res.errorCode) {
            data.errorCode = res.errorCode;
        }
        events_handler_1.default.fire('request', data);
    });
}
exports.default = requestHandler;
