// oc/dist/registry/domain/extract-package.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const targz_1 = __importDefault(require("targz"));
const util_1 = require("util");
const get_package_json_from_temp_dir_1 = __importDefault(require("./get-package-json-from-temp-dir"));
async function extractPackage(files, tarExtractMode) {
    const packageFile = files[0];
    const packagePath = path_1.default.resolve(packageFile.path);
    const packageUntarOutput = path_1.default.resolve(packageFile.path, '..', packageFile.filename.replace('.tar.gz', ''));
    const packageOutput = path_1.default.resolve(packageUntarOutput, '_package');
    const decompress = (0, util_1.promisify)(targz_1.default.decompress);
    await decompress({
        src: packagePath,
        dest: packageUntarOutput,
        tar: {
            dmode: tarExtractMode,
            fmode: tarExtractMode
        }
    });
    const packageJson = await (0, get_package_json_from_temp_dir_1.default)(packageOutput);
    return {
        outputFolder: packageOutput,
        packageJson
    };
}
exports.default = extractPackage;
