// oc/dist/cli/domain/url-parser.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.parse = void 0;
const querystring_1 = __importDefault(require("querystring"));
const url_1 = __importDefault(require("url"));
const removeFinalSlashes = function (s) {
    while (s.slice(-1) === '/') {
        s = s.slice(0, -1);
    }
    return s;
};
function parse(parsed) {
    const requestedVersion = parsed.requestVersion;
    const href = url_1.default.parse(parsed.href);
    const relativePath = removeFinalSlashes(href.pathname || '');
    const withoutVersion = removeFinalSlashes(relativePath.replace(requestedVersion, ''));
    const componentName = withoutVersion.substr(withoutVersion.lastIndexOf('/') + 1);
    const withoutComponent = removeFinalSlashes(withoutVersion.replace(componentName, ''));
    const registryUrl = href.protocol + '//' + href.host + withoutComponent + '/';
    return {
        clientHref: registryUrl + 'oc-client/client.js',
        name: componentName,
        parameters: querystring_1.default.parse(href.query || ''),
        registryUrl: registryUrl,
        version: requestedVersion
    };
}
exports.parse = parse;
