// oc/dist/cli/domain/watch.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const chokidar_1 = __importDefault(require("chokidar"));
const settings_1 = __importDefault(require("../../resources/settings"));
function watch(dirs, baseDir, changed) {
    const watcher = chokidar_1.default.watch(path_1.default.resolve(baseDir), {
        ignored: settings_1.default.filesToIgnoreOnDevWatch,
        persistent: true,
        ignoreInitial: true,
        usePolling: true
    });
    const onChange = (fileName) => {
        const componentDir = dirs.find(dir => Boolean(fileName.match(escapeRegularExpression(dir + path_1.default.sep))));
        changed(null, fileName, componentDir);
    };
    watcher
        .on('add', onChange)
        .on('change', onChange)
        .on('unlink', onChange)
        .on('error', changed);
}
exports.default = watch;
function escapeRegularExpression(text) {
    return text.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
}
